Add additional STM32WL SOCS and Seeed Studio LoRa-E5 Dev board

str4t0m opened the pull request:
This PR adds SOC configuration and device tree definitions for additional STM32WL SOCS.
Additionally support for the Seeed Studio LoRa-E5 Dev board is added [(LoRa-E5 Dev board Wiki](https://wiki.seeedstudio.com/LoRa_E5_Dev_Board/)).
This board is powered by a STM32WLE5JC soc.

Edit: After merge of #34751 all tests can now successfully be build.

str4t0m committed:
dts: stm32wl: add definitions for further stm32wl  socs.

This commit adds dt support for stm32wle4, and stm32wle5 single core,
as well as stm32wl54 dual core socs.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
boards: add seeed studio LoRa-E5 Dev board

This commit adds support for the seeed studio LoRa-E5 Dev board,
which is powered by a module based on stm32wle5jc soc.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
soc/arm: add support for additonal stm32wl socs

This commit adds support for stm32wle4xx, stm32wle5xx single core socs,
as well as stm32wl54 dual core soc.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: stm32wl set cpu2 prescaler only for STM32WL5X

To support single core stm32wlex series, cpu2 prescaler is set
only on dual core soc variants.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m commented on dts/arm/st/wl/stm32wl54Xc.dtsi:
```diff
@@ -0,0 +1,21 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <mem.h>
+#include <st/wl/stm32wl55.dtsi>
```
As stm32wlx5 socs differ from stm32wlx4 socs only with additional support for lora modulation a common include file is used.
Do you think there might any dt differences between this socs in the future, making it preferable including  a <st/wl/stm32wl54.dtsi> file instead?

erwango replied:
Let's see if @ABOSTM has more info, but I'm fine to go as is and update later if need be.

str4t0m commented:
All dependencies are merged now and all CI tests pass.

erwango commented:
@fabiobaltieri would you mind having a look ?

erwango commented on boards/arm/lora_e5_dev_board/lora_e5_dev_board.dts:
```diff
@@ -0,0 +1,177 @@
+		label = "pwr-3v3-ctrl";
+		regulator-name = "pwr-3v3-ctrl";
+		enable-gpios = <&gpioa 9 GPIO_ACTIVE_HIGH>;
+		regulator-boot-on;
```
Please add status

str4t0m replied:
Done.

erwango commented on boards/arm/lora_e5_dev_board/lora_e5_dev_board.dts:
```diff
@@ -0,0 +1,177 @@
+		label = "pwr-5v-ctrl";
+		regulator-name = "pwr-5v-ctrl";
+		enable-gpios = <&gpiob 10 GPIO_ACTIVE_HIGH>;
+		regulator-boot-on;
```
Please add explicit status

str4t0m replied:
Done.

erwango commented on boards/arm/lora_e5_dev_board/lora_e5_dev_board.dts:
```diff
@@ -0,0 +1,177 @@
+grove_i2c: &i2c2 {};
+
+&flash0 {
+	/*
+	 * For more information, see:
+	 * http://docs.zephyrproject.org/latest/guides/dts/index.html#flash-partitions
+	 */
```
Please remove this.
(we're trying to get rid of those)

str4t0m replied:
Done

erwango reviewed (changes requested):
Some small comments, otherwise LGTM

fabiobaltieri commented on boards/arm/lora_e5_dev_board/lora_e5_dev_board.dts:
```diff
@@ -0,0 +1,177 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		/* Set 2Kb of storage at the end of the 256Kb of flash */
+		storage_partition: partition@3f800 {
+			label = "storage";
+			reg = <0x0003f800 0x00000800>;
+		};
```
I'd suggest using an 8 block default for the storage partition so this can be used with the `SETTINGS` subsystem out of the box (which is convenient for saving LoRa parameters). For the nucleo board I've used:

```
/*
 * Set 16kB of storage (8x2kB pages) at the end of the 256kB of
 * flash.
 */
storage_partition: partition@3c000 {
        label = "storage";
        reg = <0x0003c000 0x00004000>;
};
```

fabiobaltieri commented on boards/arm/lora_e5_dev_board/lora_e5_dev_board.dts:
```diff
@@ -0,0 +1,177 @@
+ */
+
+/dts-v1/;
+#include <st/wl/stm32wle5Xc.dtsi>
```
Should we have a dtsi for the module? Similar to the one in https://github.com/zephyrproject-rtos/zephyr/tree/main/dts/arm/acsip.

The LoRa-E5 includes the MCU, TCXO, low frequency crystal, and radio switch (https://www.disk91.com/2021/technology/lora/lora-e5-the-seeed-studio-stm32-wl-module/). Thinking it should have the` stm32wle5Xc.dtsi` include, the clock nodes (`rcc`, `pll` and `clk_msi`) and then I can rebase the radio patch and add the radio node there with the RF switch pin mapping.

str4t0m replied:
I have added a nearly empty lora-e5.dtsi in the arm/seeed folder.
The external clock frequencies are already defined in the soc dtsi, therefore I have not defined them here again.
I have left out further clock configuration(rcc, pll) as this may often be adjusted by the user per board.

Please let me know what you think about introducing a "lora-e5" module and if you would add further definitions.

fabiobaltieri replied:
Yeah perfect! I'll rebase the lora patch on top and add the subghz node with the module setting there (like the acsip one) once this is merged. Thanks!

fabiobaltieri commented on boards/arm/lora_e5_dev_board/board.cmake:
```diff
@@ -0,0 +1,12 @@
+include(${ZEPHYR_BASE}/boards/common/pyocd.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/openocd.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/stm32cubeprogrammer.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
```
Could you add blackmagicprobe as well? Selfish comment... that's what I'm using most of the time. :-)

str4t0m replied:
For sure.

fabiobaltieri commented on boards/arm/lora_e5_dev_board/doc/lora_e5_dev_board.rst:
```diff
@@ -0,0 +1,278 @@
+  - RS-485 connector
+  - SMA-K and IPEX antenna connectors
+
+- Delivered with SMA antenna(per default IPEX connector is disconnected)
```
```suggestion
- Delivered with SMA antenna (per default IPEX connector is disconnected)
```

fabiobaltieri commented on boards/arm/lora_e5_dev_board/doc/lora_e5_dev_board.rst:
```diff
@@ -0,0 +1,278 @@
+Applications for the ``lora_e5_dev_board`` board configuration can be built the
+usual way (see :ref:`build_an_application`).
+
+Be careful to not brick the device by flashing the device before removing the
+flash protections which are initially enabled.
```
Hey more of a curiosity than anything else... how can this happens? Thought that the memory write would just fail until the device is erased and unlocked. Is there a condition that bricks it? (apart from locking it permanently)

str4t0m replied:
I missed to comment on this one after I addressed the other comments.
To be honest it is not that likely to brick the device as long as you can attach the debugger.
Mainly the option to recover using the system bootloader is no longer directly available.
I have removed this sentence, as the next sentence includes the information about the boot-0 pin not being available.

fabiobaltieri replied:
Oh I see... have you actually managed to use a custom firmware with the stock bootloader? Thought you'd had to set up a flash programming offset for that to work. If using the SWD port, the device is shipped with protection enabled, so I don't think you can write to the flash at all without doing a mass erase (that's what I did on mine), maybe worth mentioning that.

str4t0m replied:
No, I had no interest in using their bootloader, so I have not tried.
But there is hardly any information available about the bootloader, so I believe it is only supposed to provide them the possibility to upgrade their AT command firmware.

I have rewritten this paragraph to include less information about the bootloader, and instead added a hint that rdp regression must be performed.
Could you please have another look?

fabiobaltieri replied:
Yeah looking good, thanks! Is there any step worth documenting to reset the device protection to level 0 with the stlink? I did it manually with a jlink on mine but I presume they have it in their standard tool by now.

str4t0m replied:
Thanks. I have added one sentence on how the rdp level can be reset in case an stlink debugger is used.

fabiobaltieri replied:
Cool! Saw the update, that's cool, think it'd help a lot for a random user to get started. Only last thing I'd add is a note to call out that the unlock is also going to erase the device, including the preprogrammed bootloader, so the AT firmware cannot be used anymore (unclear to me if Seeed is ever going to distribute the hex files for it).

str4t0m replied:
Added a short note that the firmware will be erased:

> The RDP level 1 to RDP level 0 regression will erase the factory programmed AT
> firmware, from which seeed has neither released the source code nor a binary.

fabiobaltieri replied:
Cool! Thanks for following up on this!

str4t0m replied:
Thanks for the recommendations.

ABOSTM commented on soc/arm/st_stm32/stm32wl/Kconfig.soc:
```diff
@@ -7,7 +7,21 @@ choice
 
 endchoice
+
+config STM32WL_DUAL_CORE
```
Is this switch supposed to be configured by user, or implicitly depending on MCU used ?

Knowing that there is already a CMSIS switch `DUAL_CORE ` (for example https://github.com/zephyrproject-rtos/hal_stm32/blob/d49fb6dbe780b931eee33cce8656e87c46070fa8/stm32cube/stm32wlxx/soc/stm32wl55xx.h#L43)
I think this switch is not needed, you can use:
`#if defined(CONFIG_SOC_SERIES_STM32WLX) && defined(DUAL_CORE )`

str4t0m replied:
Thanks for this suggestion. The intention was to do it similar as stm32h7 does it and enable this at board level.
H7 uses this flag in the flash driver to enable usage of hsem for locking, in the hsem driver itself, and in soc_m7 to start the other core after the kernel was initialized.

I have had the intention to use it to start the second core, but currently I am still awaiting the hardware to be able to test this so I left this out for now.
Would it be preferable to use DUAL_CORE flag from the hal to decide the clock configuration(and in the future for flash access). Additionally later introduce a new kconfig option to start the second core?

ABOSTM replied:
Ok, on H7 side the difference between DUAL_CORE and STM32H7_DUAL_CORE is that 
* DUAL_CORE indicate that MCU is dualcore and it is directly set thanks to CMSIS based on MCU selection
*  STM32H7_DUAL_CORE indicated that Zephyr will manage the dualcore 
 (if not set, zephyr will run as a single core even if MCU is dualcore)

So I agree to mimic what is done on H7 , and keep switch CONFIG_SOC_SERIES_STM32WLX configurable by user.
But Introducing this switch suppose that zephyr is able to handle STM32WL dualcore ... which is not the case for now.  
It would require to add 
* specific M4 vs M0 configuration, DeviceTree, yaml, ...
In other world: be able to compile zephyr for M4 and for M0 (currently only M4 is supported)
* dualcore boot:  for example boot 1st on M4 which will configure the clock, while M0 is stopped, then M4 wakeup M0 , and M0 do not configure the clock ...
* lowpower enter/exit : how to manage the clock
* manage Hardware semaphore everywhere it is necessary (shared IP like RCC, PWR, ...)
* ...

str4t0m replied:
Yes agree.
I have removed the stm32wl dual core configuration option, as at this point it does not yet make sense.

erwango replied:
Sorry to be late, but I thing we should avoid direct use of `DUAL_CORE` for the reason it is unclear and easily misleading. Its usage is fine inside CubeWL, but outside (and here in zephyr) it could lead to misunderstanding as it lacks basic namespacing.
So, to the minimum, I suggest we introduce a dedicated macro coming with some context explanation.
This will work for this driver, but if we need this across drivers I'm not sure we'll be able to avoid defining a Kconfig symbol as suggested initially.

str4t0m replied:
Following the suggestion to avoid direct usage of DUAL_CORE, I am now using:

```
/* Identify stm32wl dual-core socs by symbol defined in CMSIS dev header file */
#if (defined(CONFIG_SOC_SERIES_STM32WLX) && defined(DUAL_CORE))
#define STM32WL_DUAL_CORE
#endif
```

And we can introduce a Kconfig symbol, once better better dual core support is available.
Any alternative suggestions?

ABOSTM commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -75,7 +75,7 @@
 static void config_bus_clk_init(LL_UTILS_ClkInitTypeDef *clk_init)
 {
-#if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_SOC_SERIES_STM32WLX)
+#if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32WL_DUAL_CORE)
```
See previous remark

str4t0m committed:
dts: arm: introduce seeed lora-e5 module

This commit adds the dts definitons for the seeed lora-e5 module.
Additionally I add myself as codeowner for the new dts/arm/seeed
directory.

This module packages a stm32wle5jc Sub-GHz Wireless Soc,
together with a 32MHz TCXO, a 32.768KHz crystal oscillator, and
power and RF circuitry.

With the introduction of lora support definitions for the radio
will be added in a future commit.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

ABOSTM commented on dts/arm/st/wl/stm32wle4Xb.dtsi:
```diff
@@ -0,0 +1,21 @@
+/*
```
There is no MCU stm32wle4X8 or stm32wle4Xb 
Starting with stm32wle4 , only stm32wle4cc and stm32wle4jc exist

str4t0m replied:
Hm, I see those are not listed on the ST Website, but are listed in STM32Cube and the Keil device family pack.
Do you know if they may be introduced later?

ABOSTM replied:
Ok if you found them in CubeMX, I guess it will come later.
So it is good for me to keep them

ABOSTM commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -75,7 +75,8 @@
 static void config_bus_clk_init(LL_UTILS_ClkInitTypeDef *clk_init)
 {
-#if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_SOC_SERIES_STM32WLX)
+#if defined(CONFIG_SOC_SERIES_STM32WBX) || \
+	defined(CONFIG_SOC_SERIES_STM32WLX) && defined(DUAL_CORE)
```
To avoid ambiguity, I would add some parenthesis
```suggestion
#if defined(CONFIG_SOC_SERIES_STM32WBX) || \
	(defined(CONFIG_SOC_SERIES_STM32WLX) && defined(DUAL_CORE))
```

ABOSTM commented on dts/arm/seeed/lora-e5.dtsi:
```diff
@@ -0,0 +1,12 @@
+#include <st/wl/stm32wle5jcix-pinctrl.dtsi>
+
+&clk_hse {
+	hse-bypass;
```
Missing `clock-frequency `property ?

str4t0m replied:
stm32wl.dtsi already defines the correct frequency for this series:
```
/* Expected clock-frequency on the whole series 32MHz */
clock-frequency = <DT_FREQ_M(32)>;
```
I thought about re-defining it anyways and decided initially against it.
Will add it now anyways.

fabiobaltieri commented on dts/arm/seeed/lora-e5.dtsi:
```diff
@@ -0,0 +1,17 @@
+
+&clk_lse {
+	clock-frequency = <32768>;
+}
```
Think you are missing a ; here.

str4t0m replied:
Indeed. Thanks.

ABOSTM reviewed (approved):
LGTM

str4t0m commented:
Rebased on latest upstream/main.
I believe this PR now could already be merged.
